doc/crypto/api/keys/management.rst

```diff
@@ -103,17 +397,151 @@ When creating a key, the attributes for the new key are specified in a `psa_key_
-
-    Implementations must reject an attempt to import a key of size ``0``.
+    The key data determines the key size. :code:``psa_get_key_bits(attributes)`` must either match the determined key size or be ``0``. Implementations must reject an attempt to import a key of size zero.
 
-    This function supports any output from `psa_export_key()`. Each key type in :secref:`key-types` describes the expected format of keys.
+    This function supports any output from `psa_export_key()`.
+    The default format is defined in the *Key format* section for each key type.
+    Other key formats can be imported using `psa_import_formatted_key()`.
 
-    This specification defines a single format for each key type. Implementations can optionally support other formats in addition to the standard format. It is recommended that implementations that support other formats ensure that the formats are clearly unambiguous, to minimize the risk that an invalid input is accidentally interpreted according to a different format.
+    Implementations can optionally support other formats in calls to `psa_import_key()`, in addition to the default format.
 
     .. note::
         The |API| does not support asymmetric private key objects outside of a key pair. To import a private key, the ``attributes`` must specify the corresponding key pair type. Depending on the key type, either the import format contains the public key data or the implementation will reconstruct the public key from the private key as needed.
 
+    .. admonition:: Implementation note
+
+        It is recommended that implementations that support other formats ensure that the formats are clearly unambiguous, to minimize the risk that an invalid input is accidentally interpreted according to a different format.
+
+        It is recommended that implementations reject key data if it might be erroneous, for example, if it is the wrong type or is truncated.
+
+.. function:: psa_import_formatted_key
+
+    .. summary::
+        Import a key in a specified format.
+        *   The key type can be `PSA_KEY_TYPE_NONE`. If either is nonzero, it must match the corresponding attribute of the source key.
+
+    .. param:: const psa_key_attributes_t * attributes
+        The attributes for the new key.
+
+        Depending on the specified key format, and the attributes encoded in the key data, some of the key attributes can be optional.
+
+        The following attributes are required for formats that do not specify a key type:
+
+        *   When the format does not specify a key type: the key type in ``attributes`` determines how the ``data`` buffer is interpreted.
+        *   When the format does specify a key type: if the key type in ``attributes`` has a non-default value, it must be equal to the determined key type.
+
+        The following attributes must be set for keys used in cryptographic operations:
+
+        *   The key permitted-algorithm policy, see :secref:`permitted-algorithms`.
```

Comment:
Blocker in the API design: when I'm loading a key which may be of several types, the desired algorithm policy will depend on the type. A typical example: my application can sign with either RSA-PKCS#1v1.5 or ECDSA. I want to set the algorithm policy to `PSA_ALG_RSA_PKCS1V15_SIGN_ANY` if the key turns out to be an RSA key, and to `PSA_ALG_ECDSA_ANY` if it turns out to be an ECC key.

In the Mbed TLS 3.x equivalent to this interface (pk API), where we were extending the legacy interface to bridge it with PSA, I resolved this by making it a [three-step process](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.0/docs/psa-transition.md#creating-a-psa-key-via-pk):

1. Parse the key. This creates a legacy object which has type information, but no binding policy information.
2. Construct attributes, including a policy, based on the key type. There's a helper function that handles the common cases, but the application can tweak the result or just make up its own attributes.
3. Import the key into PSA, with the policy and other attributes chosen at the previous step.

This only works because there's an intermediate stage with a key that is loaded and has type information but does not yet have a policy set. I'd rather avoid having to have this intermediate stage in the PSA API.

I'm currently working on the evolution of Mbed TLS pk API for the next major version, which will be a transition API (keeping old function names, but wrapping around PSA keys). I don't have a satisfactory solution yet in that context either.

Reply:
The obvious way is to encode, either in the `format` argument or in a separate argument, the mapping for “if the loaded key has that type then use this policy”. But that gets pretty hairy.

Reply:
It does get messy if we want to encode some form of conditional policy logic in a parameter to enable something like "if EC key then permit ECDSA, else if RSA then permit RSA-PSS, else if ....". How far should such flexibility extend?

I can envisage an API that parsed the key data to be able to report the attributes of the key that is encoded in the data, but without creating a key object. The application can then use or amend the reported key attributes before actually importing the key. If the application does not need this flexibility, it can just import the key. The overhead of this approach is parsing the key data twice, which is perhaps particularly an issue for wrapped/encrypted formats.

Reply:
In a scenario where the key format provides enough information, and the application is prepared to delegate the policy setting to the key data, do we need a permitted algorithm and usage flag value that explicitly indicates this?

For example, if the application sets all the available usage flags - then a strict "intersection with the policy in the key data" would have this effect, but what should the final key policy be if the key data has no explicitly specified policy? (presumably not 'can do anything')

Similarly, we probably want to have different ways to differeniate "no algorithm is permitted" from "I am not specifying a permitted algorithm - use the use one encoded in the AlgorithmIdentifier in the key data".